### ISSUE TITLE
Incorrect column names in as.matrix.projection()

### ISSUE THREAD

**Issue by fweber144** [stan-dev/projpred#72]
It seems like the column names of the matrix returned by `as.matrix.projection()` are not correct (at least not always). Here is a reprex: First run
```r
# Original source: https://github.com/stan-dev/projpred/blob/master/README.md
library(projpred)
library(rstanarm)
options(mc.cores = parallel::detectCores(logical = FALSE))
data("df_gaussian", package = "projpred")
df_gaussian <- df_gaussian[1:40, ]
mydat <- cbind("y" = df_gaussian$y, as.data.frame(df_gaussian$x))
mydat$mygroup <- gl(n = 8, k = floor(nrow(mydat) / 8),
                    labels = paste0("gr", seq_len(8)))
set.seed(457211)
mydat$noise <- rnorm(nrow(mydat))
mygroup_icpts_truth <- rnorm(nlevels(mydat$mygroup), sd = 6)
mygroup_V1_truth <- rnorm(nlevels(mydat$mygroup), sd = 6)
mydat$y <- mydat$y + mygroup_icpts_truth[as.numeric(mydat$mygroup)]
mydat$y <- mydat$y + mygroup_V1_truth[as.numeric(mydat$mygroup)] * mydat$V1

myfit <- stan_glmer(y ~ V1 + V2 + V3 + V4 + V5 + noise + (1 + V1 | mygroup),
                    data = mydat,
                    seed = 1140350788)

cvs <- cv_varsel(myfit)
myproj <- project(cvs,
                  nterms = length(solution_terms(cvs)),
                  ndraws = ceiling(0.05 * nrow(as.matrix(myfit))))
debugonce(projpred:::as.matrix.projection)
myproj_mat <- as.matrix(myproj)
```
and then debug the last call up to (but excluding) line
https://github.com/stan-dev/projpred/blob/04326a1ceaedccf4883a1a6fb6c04b77d186c08d/R/methods.R#L835
If you then compare `x$solution_terms` and `colnames(res)`, then you should see that the ordering of the terms in `x$solution_terms` doesn't match the ordering in `colnames(res)`:
```r
# x$solution_terms:
# [1] "(1 | mygroup)"  "V1"             "(V1 | mygroup)" "V3"             "V5"             "noise"          "V4"            
# [8] "V2"  
# colnames(res):
# [1] "(Intercept)"         "V1"                  "V3"                  "V5"                  "noise"              
# [6] "V4"                  "V2"                  "mygroup.(Intercept)" "mygroup.V1" 
```
Therefore, `colnames(res) <- c("Intercept", x$solution_terms)` causes incorrect column names for `res`. The issue seems to be fixed (at least as far as I have tested it) by replacing line
https://github.com/stan-dev/projpred/blob/04326a1ceaedccf4883a1a6fb6c04b77d186c08d/R/methods.R#L835
by
```r
colnames(res) <- gsub("\\(Intercept\\)", "Intercept", colnames(res))
```
I'll create a pull request for that.

Session info (note that I'm using projpred's CRAN version, but the issue probably also occurs for projpred's `master` and `develop` GitHub branches since the line mentioned above hasn't changed in those GitHub branches):
```
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
[...]

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] rstanarm_2.21.2 Rcpp_1.0.6      projpred_2.0.2 

loaded via a namespace (and not attached):
 [1] nlme_3.1-152            matrixStats_0.58.0      xts_0.12.1              threejs_0.3.3           splines2_0.4.1         
 [6] rstan_2.26.0.9000       numDeriv_2016.8-1.1     tools_4.0.3             R6_2.5.0                DT_0.17                
[11] DBI_1.1.1               mgcv_1.8-33             colorspace_2.0-0        tidyselect_1.1.0        gridExtra_2.3          
[16] prettyunits_1.1.1       processx_3.4.5          curl_4.3                compiler_4.0.3          cli_2.3.0              
[21] shinyjs_2.0.0           colourpicker_1.1.0      scales_1.1.1            dygraphs_1.1.1.6        ggridges_0.5.3         
[26] callr_3.5.1             stringr_1.4.0           digest_0.6.27           StanHeaders_2.26.0.9000 minqa_1.2.4            
[31] rmarkdown_2.6           base64enc_0.1-3         pkgconfig_2.0.3         htmltools_0.5.1.1       lme4_1.1-26            
[36] fastmap_1.1.0           htmlwidgets_1.5.3       rlang_0.4.10            shiny_1.6.0             generics_0.1.0         
[41] zoo_1.8-8               jsonlite_1.7.2          crosstalk_1.1.1         gtools_3.8.2            dplyr_1.0.4            
[46] inline_0.3.17           magrittr_2.0.1          loo_2.4.1.9000          bayesplot_1.8.0         Matrix_1.3-2           
[51] munsell_0.5.0           lifecycle_0.2.0         stringi_1.5.3           MASS_7.3-53             pkgbuild_1.2.0         
[56] plyr_1.8.6              grid_4.0.3              parallel_4.0.3          promises_1.1.1          crayon_1.4.0           
[61] miniUI_0.1.1.1          lattice_0.20-41         splines_4.0.3           knitr_1.31              ps_1.5.0               
[66] pillar_1.4.7            igraph_1.2.6            boot_1.3-25             markdown_1.1            shinystan_2.5.0        
[71] reshape2_1.4.4          codetools_0.2-18        stats4_4.0.3            rstantools_2.1.1        glue_1.4.2             
[76] evaluate_0.14           V8_3.4.0                RcppParallel_5.0.2      vctrs_0.3.6             nloptr_1.2.2.2         
[81] httpuv_1.5.5            gtable_0.3.0            purrr_0.3.4             assertthat_0.2.1        ggplot2_3.3.3          
[86] xfun_0.20               mime_0.9                xtable_1.8-4            later_1.1.0.1           rsconnect_0.8.16       
[91] survival_3.2-7          tibble_3.0.6            shinythemes_1.2.0       optimx_2020-4.2         gamm4_0.2-6            
[96] statmod_1.4.35          ellipsis_0.3.1  
```

**Comment by AlejandroCatalina** [comment]
So this is a bit tricky and I also realised that some models returned different names. It seems to happen with both multilevel models and models with contrasts, as the underlying projections have names for the contrasts. 

I think your solution can partially solve the issue in some models but we still have the thing of fixing the group names, as in your example we recover `my_group.(Intercept)` and so on while we ideally want `(Intercept | my_group)`, etc. I'll reply in your PR with this in case you want to address that as well?

**Comment by fweber144** [comment]
I also thought about keeping the old column names style (from `x$solution_terms`, but re-ordered), but I have two other issues that I wanted to submit afterwards and whose fixes make sense with this new style of column names. I'll create PRs for those two new issues and then it might be easier to discuss the naming of the columns.

Thanks for the hint concerning contrasts. I will have to see whether my proposed fix has to be adopted for contrasts.

**Comment by AlejandroCatalina** [comment]
What happens in the case of contrasts is that the projection matrix has more columns than names in x$solution_terms, so in that case we may be better off just keeping the final columns in the matrix for the contrasts.

Best,
Alejandro

**Comment by fweber144** [comment]
Yes, the same holds for my two other issues.

**Comment by fweber144** [comment]
> I also thought about keeping the old column names style (from `x$solution_terms`, but re-ordered), but I have two other issues that I wanted to submit afterwards and whose fixes make sense with this new style of column names. I'll create PRs for those two new issues and then it might be easier to discuss the naming of the columns.

Here are these two new PRs: #74 and #75.

**Comment by fweber144** [comment]
Fixed by merged PR #75.
